src/init.cpp

```diff
@@ -1444,8 +1445,18 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     nTotalCache = std::max(nTotalCache, nMinDbCache << 20); // total cache cannot be less than nMinDbCache
     nTotalCache = std::min(nTotalCache, nMaxDbCache << 20); // total cache cannot be greated than nMaxDbcache
     int64_t nBlockTreeDBCache = nTotalCache / 8;
-    if (nBlockTreeDBCache > (1 << 21) && !GetBoolArg("-txindex", false))
-        nBlockTreeDBCache = (1 << 21); // block tree db cache shouldn't be larger than 2 MiB
+    if (GetBoolArg("-addressindex", false)) {
+        if (!GetBoolArg("-txindex", false)) {
+            return InitError(_("-addressindex requires -txindex."));
+        }
+        // enable 3/4 of the cache if addressindex is enabled
+        nBlockTreeDBCache = nTotalCache * 3 / 4;
+
+    } else {
+        if (nBlockTreeDBCache > (1 << 21) && !GetBoolArg("-txindex", false)) {
+            nBlockTreeDBCache = (1 << 21); // block tree db cache shouldn't be larger than 2 MiB
+        }
+    }
```

Comment:
source for above: https://github.com/bitpay/bitcoin/commit/c91d78b578a8700a45be936cb5bb0931df8f4b87#diff-c865a8939105e6350a50af02766291b7R1233 except I added a requirement that the `-addressindex` option requires the `-txindex` option, which I'm almost certain I came to understand is true (but I can't provide a reference, sorry). Note also that I'm not adding db compression or maxopenfiles changes yet, because I'm not yet convinced they are needed, and also to keep this PR reasonably small. We can revisit later.

Reply:
Is there anything in the bitpay upstream code which confirms the requirement for `-txindex`?

Reply:
See https://github.com/bitpay/bitcore-node/blob/master/docs/upgrade.md (find in page `txindex`), but you probably noticed that the test I added at https://github.com/zcash/zcash/pull/3837/files#diff-c865a8939105e6350a50af02766291b7R1449 isn't in upstream (either hackworks or bitpay). I added it because I thought it would be an improvement to require `-txindex`, but should I remove that test so that we match upstream? That change would be a separate commit (too late to force-push it).